# Loan Data Set Exploration

## Importing required libraries

In [39]:
import pandas as pd
import numpy as np
import scipy.stats

import matplotlib
import matplotlib.pyplot as pp
import seaborn as sns

import pandas.plotting

from IPython import display
from ipywidgets import interact, widgets

%matplotlib inline
sns.set_theme(context='notebook', style='darkgrid')

In [40]:
loan = pd.read_csv('loan.csv', dtype={'Current Credit Balance': 'int64'})

C:\Users\imran\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3441: DtypeWarning: Columns (16) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [41]:
loan.head()

Loan ID                           Customer ID  \
0  000025bb-5694-4cff-b17d-192b1a98ba44  5ebc8bb1-5eb9-4404-b11b-a6eebc401a19   
1  00002c49-3a29-4bd4-8f67-c8f8fbc1048c  927b388d-2e01-423f-a8dc-f7e42d668f46   
2  00002d89-27f3-409b-aa76-90834f359a65  defce609-c631-447d-aad6-1270615e89c4   
3  00005222-b4d8-45a4-ad8c-186057e24233  070bcecb-aae7-4485-a26a-e0403e7bb6c5   
4  0000757f-a121-41ed-b17b-162e76647c1f  dde79588-12f0-4811-bab0-e2b07f633fcd   

  Loan Status  Current Loan Amount        Term  Credit Score  \
0  Fully Paid                11520  Short Term         741.0   
1  Fully Paid                 3441  Short Term         734.0   
2  Fully Paid                21029  Short Term         747.0   
3  Fully Paid                18743  Short Term         747.0   
4  Fully Paid                11731  Short Term         746.0   

  Years in current job Home Ownership  Annual Income             Purpose  \
0            10+ years  Home Mortgage        33694.0  Debt Consolidation   
1              4 years  Home Mortgage        42269.0               other   
2            10+ years  Home Mortgage        90126.0  Debt Consolidation   
3            10+ years       Own Home        38072.0  Debt Consolidation   
4              4 years           Rent        50025.0  Debt Consolidation   

  Monthly Debt  Years of Credit History  Months since last delinquent  \
0      $584.03                     12.3                          41.0   
1   $1,106.04                      26.3                           NaN   
2   $1,321.85                      28.8                           NaN   
3      $751.92                     26.2                           NaN   
4      $355.18                     11.5                           NaN   

   Number of Open Accounts  Number of Credit Problems  Current Credit Balance  \
0                       10                          0                    6760   
1                       17                          0                    6262   
2                        5                          0                   20967   
3                        9                          0                   22529   
4                       12                          0                   17391   

  Maximum Open Credit  Bankruptcies  Tax Liens  
0               16056           0.0        0.0  
1               19149           0.0        0.0  
2               28335           0.0        0.0  
3               43915           0.0        0.0  
4               37081           0.0        0.0

In [42]:
loan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 256984 entries, 0 to 256983
Data columns (total 19 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Loan ID                       256984 non-null  object 
 1   Customer ID                   256984 non-null  object 
 2   Loan Status                   256984 non-null  object 
 3   Current Loan Amount           256984 non-null  int64  
 4   Term                          256984 non-null  object 
 5   Credit Score                  195308 non-null  float64
 6   Years in current job          245508 non-null  object 
 7   Home Ownership                256984 non-null  object 
 8   Annual Income                 195308 non-null  float64
 9   Purpose                       256984 non-null  object 
 10  Monthly Debt                  256984 non-null  object 
 11  Years of Credit History       256984 non-null  float64
 12  Months since last delinquent  116601 non-nul

In [43]:
loan.describe()

Current Loan Amount   Credit Score  Annual Income  \
count         2.569840e+05  195308.000000   1.953080e+05   
mean          1.371331e+07    1251.116099   7.195272e+04   
std           3.438131e+07    1762.016848   5.887757e+04   
min           5.050000e+02     585.000000   0.000000e+00   
25%           8.299000e+03     714.000000   4.432100e+04   
50%           1.429800e+04     733.000000   6.124200e+04   
75%           2.436700e+04     744.000000   8.646200e+04   
max           1.000000e+08    7510.000000   8.713547e+06   

       Years of Credit History  Months since last delinquent  \
count            256984.000000                 116601.000000   
mean                 18.290195                     34.881450   
std                   7.075747                     21.854165   
min                   3.400000                      0.000000   
25%                  13.500000                     16.000000   
50%                  17.000000                     32.000000   
75%                  21.700000                     51.000000   
max                  70.500000                    176.000000   

       Number of Open Accounts  Number of Credit Problems  \
count            256984.000000              256984.000000   
mean                 11.106267                   0.156628   
std                   4.982982                   0.460731   
min                   0.000000                   0.000000   
25%                   8.000000                   0.000000   
50%                  10.000000                   0.000000   
75%                  14.000000                   0.000000   
max                  76.000000                  11.000000   

       Current Credit Balance   Bankruptcies      Tax Liens  
count            2.569840e+05  256455.000000  256961.000000  
mean             1.540656e+04       0.110316       0.027203  
std              1.966506e+04       0.336229       0.245950  
min              0.000000e+00       0.000000       0.000000  
25%              5.974000e+03       0.000000       0.000000  
50%              1.107800e+04       0.000000       0.000000  
75%              1.931900e+04       0.000000       0.000000  
max              1.731412e+06       7.000000      11.000000

## Step 1: Cleaning the Dataframe

### Credit Score: Dealing with missing values

In [44]:
loan.isna().sum()

Loan ID                              0
Customer ID                          0
Loan Status                          0
Current Loan Amount                  0
Term                                 0
Credit Score                     61676
Years in current job             11476
Home Ownership                       0
Annual Income                    61676
Purpose                              0
Monthly Debt                         0
Years of Credit History              0
Months since last delinquent    140383
Number of Open Accounts              0
Number of Credit Problems            0
Current Credit Balance               0
Maximum Open Credit                  0
Bankruptcies                       529
Tax Liens                           23
dtype: int64

* Credit score: 61676 missing entries, that's a significant propportion of the data. Dropping the rows is not a good option in this case. We will replace those with the mean credit score.

In [45]:
loan['Credit Score'] = loan['Credit Score'].fillna(loan['Credit Score'].mean())


In [46]:
loan['Credit Score'].isna().sum()

0

### Years in current job

* There are 11476 missing entries
* We can change it to integer, by dropping the '+ years' part of the entry and any other invalud string

In [92]:
loan['Years in current job'] = loan['Years in current job'].str.strip(r'+ years').replace('< 1', '1').fillna(0)

In [93]:
loan['Years in current job'] = loan['Years in current job'].astype(int)

In [96]:
loan['Years in current job'].describe()

count    256984.000000
mean          5.607295
std           3.746434
min           0.000000
25%           2.000000
50%           6.000000
75%          10.000000
max          10.000000
Name: Years in current job, dtype: float64

### Annual Income
* There are 61676 missing entries, the same number of missing entries as Credit Score. Similaliry, we can fill these with the mean.

In [100]:
loan['Annual Income'] = loan['Annual Income'].fillna(loan['Annual Income'].mean())

In [101]:
loan['Annual Income'].isnull().sum()

0

### Months since last delinquent 

* 140383 nan entries for Months since last delinquent.
* Looking at the data, it can be result of not having any late payments, therefore we can replace it with 0.

In [106]:
loan['Months since last delinquent']= loan['Months since last delinquent'].fillna(0)

In [107]:
loan['Months since last delinquent'].isnull().sum()

0

### Bankruptcies

* 529 missing entries, we can fill it with 0 as the mean is 0.11

In [109]:
loan['Bankruptcies'] = loan['Bankruptcies'].fillna(0)

### Tax Liens
* Only 29 missing entries, with mean of 0.02, we will fill it with 0

In [111]:
loan['Tax Liens'] = loan['Tax Liens'].fillna(0)

In [112]:
loan.isna().sum()

Loan ID                         0
Customer ID                     0
Loan Status                     0
Current Loan Amount             0
Term                            0
Credit Score                    0
Years in current job            0
Home Ownership                  0
Annual Income                   0
Purpose                         0
Monthly Debt                    0
Years of Credit History         0
Months since last delinquent    0
Number of Open Accounts         0
Number of Credit Problems       0
Current Credit Balance          0
Maximum Open Credit             0
Bankruptcies                    0
Tax Liens                       0
dtype: int64

* We have completed filling all the missing entries